File: backend/src/main/java/com/hyperformancelabs/backend/service/CustomerService.java
package com.hyperformancelabs.backend.service;

import com.hyperformancelabs.backend.dto.CustomerResponseDTO;
import com.hyperformancelabs.backend.dto.GetAllCustomer;
import com.hyperformancelabs.backend.dto.LoginRequest;
import com.hyperformancelabs.backend.dto.RegisterRequest;
import com.hyperformancelabs.backend.model.Customer;

import java.util.List;
import java.util.Map;

public interface CustomerService {
    //
//    List<GetAllCustomer> getAllCustomers();
     CustomerResponseDTO register(RegisterRequest request);
//    Customer register(RegisterRequest request);

    Customer login(LoginRequest request);

    String loginAndGenerateToken(LoginRequest request);

    Customer getCustomerByUsername(String username);

    void changePassword(String username, String oldPassword, String newPassword);
    
    // Phương thức mới để lấy số lượng khách hàng mới trong khoảng thời gian
    Integer getNewCustomersCountByDateRange(String startDate, String endDate);
    
    /**
     * Lấy số lượng khách hàng mới trong kỳ trước với cùng độ dài thời gian
     * @param startDate Ngày bắt đầu của kỳ hiện tại (định dạng dd/MM/yyyy)
     * @param endDate Ngày kết thúc của kỳ hiện tại (định dạng dd/MM/yyyy)
     * @return Số lượng khách hàng mới trong kỳ trước
     */
    Integer getNewCustomersCountInPreviousPeriod(String startDate, String endDate);
    
    /**
     * Lấy số lượng khách hàng mới và phần trăm thay đổi so với kỳ trước
     * @param startDate Ngày bắt đầu của kỳ hiện tại (định dạng dd/MM/yyyy)
     * @param endDate Ngày kết thúc của kỳ hiện tại (định dạng dd/MM/yyyy)
     * @return Map chứa số lượng khách hàng mới và phần trăm thay đổi
     */
    Map<String, Object> getNewCustomersCountWithPercentChange(String startDate, String endDate);

    // NEW METHODS: admin listing & search
    com.hyperformancelabs.backend.dto.CustomerListResponse getAllCustomers(int page, int size, String sortBy, org.springframework.data.domain.Sort.Direction sortDirection, java.util.Map<String, String> filters);

    com.hyperformancelabs.backend.dto.CustomerListResponse searchCustomers(String keyword, int page, int size);

    com.hyperformancelabs.backend.dto.CustomerDTO getCustomerById(Integer customerId);
    com.hyperformancelabs.backend.dto.CustomerDTO createCustomer(com.hyperformancelabs.backend.dto.CustomerCreateRequest request);
    com.hyperformancelabs.backend.dto.CustomerDTO updateCustomer(Integer customerId, com.hyperformancelabs.backend.dto.CustomerUpdateRequest request);
    void deleteCustomer(Integer customerId);

    com.hyperformancelabs.backend.dto.CustomerDTO updateCustomerStatus(Integer customerId, String status);
    com.hyperformancelabs.backend.dto.CustomerDTO updateCustomerRating(Integer customerId, Integer rating);
    com.hyperformancelabs.backend.dto.CustomerDTO adjustLoyaltyPoints(Integer customerId, Integer delta);

    java.util.List<com.hyperformancelabs.backend.dto.CustomerPaymentMethodDTO> getPaymentMethods(Integer customerId);
    com.hyperformancelabs.backend.dto.CustomerPaymentMethodDTO addPaymentMethod(Integer customerId, com.hyperformancelabs.backend.dto.CustomerPaymentMethodCreateRequest request);
    com.hyperformancelabs.backend.dto.CustomerPaymentMethodDTO updatePaymentMethod(Integer customerId, Integer cpmId, com.hyperformancelabs.backend.dto.CustomerPaymentMethodUpdateRequest request);
    void setDefaultPaymentMethod(Integer customerId, Integer cpmId);
    void deletePaymentMethod(Integer customerId, Integer cpmId);

    com.hyperformancelabs.backend.dto.CartListResponse getCarts(Integer customerId, String status, int page, int size);
    com.hyperformancelabs.backend.dto.CartDTO getCartDetail(Integer customerId, Integer cartId);

    java.util.List<com.hyperformancelabs.backend.dto.WishlistDTO> getWishlist(Integer customerId);
    void deleteWishlistItem(Integer customerId, Integer wishlistId);

    java.util.List<com.hyperformancelabs.backend.dto.ConversationDTO> getConversations(Integer customerId);
    com.hyperformancelabs.backend.dto.ConversationDetailDTO getConversationDetail(Integer customerId, Integer convId);
}

File: admin/src/services/customerService.js
import apiClient from './apiClient';

// Remove empty-string/null fields to satisfy backend validation
const cleanPayload = (obj) => {
  const cleaned = {};
  Object.entries(obj).forEach(([k, v]) => {
    if (v !== '' && v !== null && v !== undefined) cleaned[k] = v;
  });
  return cleaned;
};

const customerService = {
  // Customer CRUD
  getAllCustomers: async (params = {}) => {
    const response = await apiClient.get('/emp/customers', { params });
    return response.data;
  },
  searchCustomers: async (keyword, page = 0, size = 10) => {
    const response = await apiClient.get('/emp/customers/search', {
      params: { keyword, page, size }
    });
    return response.data;
  },
  getCustomerById: async (customerId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}`);
    return response.data;
  },
  createCustomer: async (customerData) => {
    const response = await apiClient.post('/emp/customers', cleanPayload(customerData));
    return response.data;
  },
  updateCustomer: async (customerId, customerData) => {
    const response = await apiClient.put(`/emp/customers/${customerId}`, cleanPayload(customerData));
    return response.data;
  },
  deleteCustomer: async (customerId) => {
    const response = await apiClient.delete(`/emp/customers/${customerId}`);
    return response.data;
  },
  updateCustomerStatus: async (customerId, status) => {
    const response = await apiClient.patch(`/emp/customers/${customerId}/status`, { status });
    return response.data;
  },
  updateCustomerRating: async (customerId, rating) => {
    const response = await apiClient.patch(`/emp/customers/${customerId}/rating`, { rating });
    return response.data;
  },
  adjustLoyaltyPoints: async (customerId, delta) => {
    const response = await apiClient.patch(`/emp/customers/${customerId}/loyalty-points`, { delta });
    return response.data;
  },

  // Payment Methods
  getPaymentMethods: async (customerId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}/payment-methods`);
    return response.data;
  },
  addPaymentMethod: async (customerId, data) => {
    const response = await apiClient.post(`/emp/customers/${customerId}/payment-methods`, data);
    return response.data;
  },
  updatePaymentMethod: async (customerId, cpmId, data) => {
    const response = await apiClient.put(`/emp/customers/${customerId}/payment-methods/${cpmId}`, data);
    return response.data;
  },
  setDefaultPaymentMethod: async (customerId, cpmId) => {
    const response = await apiClient.patch(`/emp/customers/${customerId}/payment-methods/${cpmId}/default`);
    return response.data;
  },
  deletePaymentMethod: async (customerId, cpmId) => {
    const response = await apiClient.delete(`/emp/customers/${customerId}/payment-methods/${cpmId}`);
    return response.data;
  },

  // Orders
  getCustomerOrders: async (customerId, params = {}) => {
    const response = await apiClient.get('/emp/orders', { params: { customerId, ...params } });
    return response.data;
  },
  getOrderDetail: async (orderId) => {
    const response = await apiClient.get(`/emp/orders/${orderId}`);
    return response.data;
  },

  // Carts
  getCustomerCarts: async (customerId, status = null, page = 0, size = 10) => {
    const query = { page, size };
    if (status) query.status = status;
    const response = await apiClient.get(`/emp/customers/${customerId}/carts`, { params: query });
    return response.data;
  },
  getCartDetail: async (customerId, cartId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}/carts/${cartId}`);
    return response.data;
  },

  // Wishlist
  getCustomerWishlist: async (customerId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}/wishlist`);
    return response.data;
  },
  deleteWishlistItem: async (customerId, wishlistId) => {
    const response = await apiClient.delete(`/emp/customers/${customerId}/wishlist/${wishlistId}`);
    return response.data;
  },

  // Conversations
  getCustomerConversations: async (customerId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}/conversations`);
    return response.data;
  },
  getConversationDetail: async (customerId, convId) => {
    const response = await apiClient.get(`/emp/customers/${customerId}/conversations/${convId}`);
    return response.data;
  }
};

export default customerService; 